### ISSUE TITLE
dasgoclient does not return result for `file dataset=x run=y lumi=z site=s`

### ISSUE THREAD

**Issue by slava77** [dmwm/das2go#34]
the following works and gives one file
```
dasgoclient --limit 0 --query 'file dataset=/HIMinBiasUPC/HIRun2011-v1/RAW run=182124 lumi=40'
```
the site query for the file shows that the file is at `T2_CH_CERN`.

However a direct combination with the site does not return anything
```
dasgoclient --limit 0 --query 'file dataset=/HIMinBiasUPC/HIRun2011-v1/RAW run=182124 lumi=40 site=T2_CH_CERN'
```

**Comment by vkuznet** [comment]
that's because there is no x-API between DBS and Rucio for that. The dataset/run/lumi are part of DBS and site is part of Rucio. DAS works with existing APIs. The x-API calls are special case which must be implemented on individual basis. Everything is fine from DAS point of view here. The former query only queries DBS, while for latter there is no existing API.

**Comment by slava77** [comment]
is `run=182124 site=T2_CH_CERN' a part of Rucio then or is it an example of x-API that actually works?

**Comment by vkuznet** [comment]
yes, `run, site` pair for dataset and block look-up is an example of x-API usage which I implemented individually, i.e. DAS supports queries like `file dataset=/a/b/c run=123 site=XYZ` and `file block=/a/b/c#123 run=123 site=XYZ`, also special case covers these queries `site dataset=/a/b/c` and `site block=/a/b/c#123`. That's it.

Said that, it is possible to implement your use case too, but I need to know how often it will be used and what is a **latency** effect is (I hope you understand that APIs queires DB and amount information which may flaw between x-API calls may vary, as such the query should be relatively short otherwise it may be cut off by cmsweb timeout threshold which is currently 5min).

**Comment by vkuznet** [comment]
may be I was not clear, `run=182124 site=T2_CH_CERN'  is an example of x-API since run is property of DBS and site is property of Rucio. In general, Rucio only knows about file/dataset/block/site, while DBS knows everything except site.

**Comment by slava77** [comment]
about frequency, at present I was going to add it to the query in the issue description, which is a part of the PR/IB tests. So, this will show up perhaps 100 times a day. But if caching is in place I could guess that the full search will be quick. OTOH, perhaps I should be looking for another solution.

The problem at hand is rather narrow: there is a fairly non-reproducible query of `file dataset=/HIMinBiasUPC/HIRun2011-v1/RAW run=182124 site=T2_CH_CERN`, which is used in IB/PR tests in CMSSW; the dasgoclient is fairly frequently returning a different set of files. This run data is all in `/HIMinBiasUPC/HIRun2011-v1/RAW#6d323a90-138b-11e1-b723-003048caaace` block, which has 301 files and I was thinking of narrowing down the set to make the query more reproducible. If we can find a solution for a repeatable query result, at least in my specific case we can replace this issue requesting the addition of `lumi` mixed with `site` with something else.

**Comment by vkuznet** [comment]
now, when you provide a query which you struggle with I can see why. the `files dataset/run/site` query is x-API one, as such what it does it first fetch all files from DBS for dataset/run pair, then it queries Rucio with files/site to see which one belong to a site. Everything runs concurrently to make query run fast, but all of those are HTTP calls to external service which (as you may guess) mail fail to process in time of original HTTP request. I bet it is is an HTTP timeout issue somewhere. You may run dasgoclient with `--verbose 3` option to see which HTTP calls it is doing and in fact you can just use those URLs to run curl calls by yourself. If some queries timeout (or fail) then you may get inconsistent results. I don't know **WHY** do you need to run this query in IB, but I think you should split query into two parts:
- run `file dataset=/HIMinBiasUPC/HIRun2011-v1/RAW run=182124` first and get list of files
- then run one or two queries like `site file=abc` to get site info
Doing this way will be much more reliable then in one go, and it will give you the same info.

**Comment by slava77** [comment]
> * run `file dataset=/HIMinBiasUPC/HIRun2011-v1/RAW run=182124` first and get list of files
> 
> * then run one or two queries like `site file=abc` to get site info

is there a way to pass a list of files to `dasgoclient --query ` or do I need to write a loop?

**Comment by vkuznet** [comment]
you need to write a loop. The list of files is what current dasgoclient is doing when you use query `file dataset=/a/b/c run=123 site=XYZ`, i.e. first dasgoclient get list of files and then it send them concurrently to Rucio. So if number of files is large and Rucio performance is poor nothing I can do about it in a single HTTP client request. The loop will run longer but will give you more control.
